### ISSUE TITLE
Where is the "keyboard shortcuts" feature?

### ISSUE THREAD

**Issue by fezhengjin** [dexterleng/homerow#6]
**Describe the bug**

I don't understand. Where is the "keyboard shortcuts" feature (CTRL+F)? I prefer the previous interactive design.

**Screenshots / video**

![image](https://user-images.githubusercontent.com/13692084/195573479-714c7a35-8a05-43f3-8d84-141f6e55856a.png)

**Comment by sunho-lee** [comment]
I don't understand too

Homerow became a new app after 0.4 version...

You can migrate [shortcat app](https://shortcat.app/)

**Comment by dexterleng** [comment]
@fezhengjin @sunho-lee @Acring

I did a short write-up on why I decided to rebuild Homerow w/ a search-by-text workflow: https://dexterleng.notion.site/To-Vimac-and-Homerow-users-a7fe71a33b24467780800d21fd26b4cf

IMO there is no one winner - hints are better for dynamic content (web), search-by-text is better for static content. It just that search-by-text is better for the apps I currently use, so I decided to start the rebuild with search-by-text feature first.

I will explore either a hybrid workflow or simply providing two modes (search by text and hints). The first alpha release of Homerow just came out this week, so hope you can be patient!

**Comment by eproxus** [comment]
Some notes after trying the original Homerow, Shortcat and the new Homerow:

* The original Homerow still has the best performing link hint mode. Shortcat takes much longer to show the hints (even when set to _Immediately_). Shortcat also shows an annoying popup on top of the screen which can't be disabled (this is true for the new Homerow too, but it is much smaller).
* The original Homerow requires the least interaction to perform a click. Both the new Homerow and Shortcat requires at least an additional <kbd>⏎</kbd> press to activate. I do like the right/middle-click modifiers better in Shortcat. The ideal would be the hint mode from the original Homerow app but with modifiers, e.g. pressing <kbd>⌘</kbd> while finishing typing a hint would right click etc.
* The search mode might be useful in some cases, but for me there are many cases where it is not usable at all. For example, what should I type to select the 3rd Hacker News tab in this example?
   <img width="411" alt="Screen Shot 2022-10-18 at 14 49 58" src="https://user-images.githubusercontent.com/112878/196434088-381d8e64-0867-4a7c-8210-dece5a693101.png">
   Impossible to know without guessing.

**Comment by zsmatrix62** [comment]
I prefer the labeling feature also just like those in Vimac as it requires less key strokes. the new version Homerow has adv 

> @fezhengjin @sunho-lee @Acring
> 
> I did a short write-up on why I decided to rebuild Homerow w/ a search-by-text workflow: https://dexterleng.notion.site/To-Vimac-and-Homerow-users-a7fe71a33b24467780800d21fd26b4cf
> 
> IMO there is no one winner - hints are better for dynamic content (web), search-by-text is better for static content. It just that search-by-text is better for the apps I currently use, so I decided to start the rebuild with search-by-text feature first.
> 
> I will explore either a hybrid workflow or simply providing two modes (search by text and hints). The first alpha release of Homerow just came out this week, so hope you can be patient! Totally understand if search-by-text doesn't suit your workflow.
> 
> I can generate a license key for the old Homerow if you purchase a new license, though please note I probably won't maintain it outside of critical bugs since the codebase is so bad.

Looking forward to this new version of combination and I must purchase if the features in Vimac which is now missing in Homerow. Cheers .. subscribing this subject for notifications.

**Comment by brettflorio** [comment]
I understand how frustrating it can be to work on an old and messy codebase, but @dexterleng, could you clarify how the new Homerow would/should work on elements like textareas or other UI elements that don't actually have any text associated with them?

The reason the previous app was so brilliant (imo) is that it made otherwise mouse-only UI elements clickable via the keyboard. For instance, focusing on the text input of a thread in Slack; clicking the `x` to close the download bar at the bottom of Chrome. It seems like this new approach makes it impossible to use the keyboard to click on elements like that? I understand the Tutor might allow me to use the mouse to try to figure out how I can then go back to the keyboard, but using the mouse is (I think) what we're using Homerow to _avoid_.

**Comment by choonkeat** [comment]
> I do like the right/middle-click modifiers better in Shortcat. The ideal would be the hint mode from the original Homerow app but with modifiers, e.g. pressing ⌘ while finishing typing a hint would right click etc.

I was like Yes! I'd like that, but then vaguely remembered there's a way to do right-click with Homerow/Vimac. Could never remember it (I've been right clicking with a real mouse 🙈) and it isn't mention anywhere..

> the workflow [of right click] has changed; now it works as a prefix:
> `r<label>` to right click
> `c<label>` to command click
> `d<label>` to double click

Let me commit into memory 😅 and use it henceforth, but I predict I'll enjoy this better than modifiers

**Comment by aaronrolls** [comment]
I am quite disappointed with the change. I hope the features that were removed are added back. Would it not have been better to release a new app with these features rather than removing the features people already used?

Currently the app is a worse version of https://Wooshy.app and the dev of Wooshy is very quick to add new features so it might be hard to compete. 

Please release an older version of Homerow.

**Comment by vzakanj** [comment]
I'm also more accustomed to the previous way of using hints since it makes my daily workflow faster, so I'm sticking with version 0.3.8 for now.

That being said, I've tried out the new version and the search approach has its merits for certain apps.
What I'd like to see in the long run is support for both modes, allowing me to activate them via separate shortcuts.

In terms of the search hybrid mode, my suggestion would be to show the hints for all elements at the start and then progressively filter them out as the search term is typed in - AceJump (https://github.com/acejump/AceJump) does this for JetBrains editors and it really works well for my use-cases.

**Comment by eret9616** [comment]
wish there wil be a new version homerow, combines shortcut feature but faster and better

**Comment by dexterleng** [comment]
I heard all your feedback and decided to give a hybrid workflow a shot 😄. Would love to hear your thoughts before I release it into the wild.

![CleanShot 2022-11-18 at 21 24 42](https://user-images.githubusercontent.com/34204380/202715249-fb818409-01e3-4081-ad58-0c3fcd9243d4.png)

## Download

[homerow-jumpers-preview.app.zip](https://github.com/dexterleng/homerow/files/10041595/homerow-jumpers-preview.app.zip)

## Changes

- Dynamically generated shortcut aka Jumpers are now shown:
    - Hold `Shift` while typing the Jumper's label to focus the element.
    - Targets are no longer shown for non-focused elements.
- Changes wildcard search:
    - You can now use `;` to perform wildcard search.
    - Added an option (enabled by default) to perform wildcard search when nothing is typed.
    - Typing wildcard characters (`*` and `;`) will auto-clear non-wildcard input and vice-versa.
- Changes to the Tutor:
    - Tutor is now automatically enabled when you move your cursor, and disabled when typing.
    - Disabled the popover's animation

cc @fezhengjin @sunho-lee @eproxus @brettflorio @aaronrolls @vzakanj @eret9616 @choonkeat

**Comment by brettflorio** [comment]
@dexterleng Thanks for taking feedback into account! At present, it seems to work great as far as making things visible and accessible, but would/could you add an option to allow the jumpers to be used without shift+jumper then pressing return? Compared to the old, where an old "jumper" would be a single keypress (or maybe two or three), the new approach requires a shift press+hold and an enter press on top of every action. Better than nothing, but still doubling or tripling the amount of keypresses needed compared to the old.

In other words, I'd rather it defaults keypresses to jumpers (without the need to hit the enter key again), and if I want to use the "search" mode, I could hit another key to switch to that. Best of both worlds and everybody's happy? :)

**Comment by lightmotive** [comment]
@dexterleng great work on the new release.

**What I appreciate**

- Changes to the Tutor: very convenient and intuitive to enable it upon moving the mouse.
- Changes to wildcard search: UX is improved imo by enabling wildcard search by default without showing every possible target.
- Jumpers: as others have noticed, this is essential for some apps.
- Hitting enter to click: to me, this is more intuitive and safe without adding too much overhead.
  - I like seeing what I'll click after I type and before I hit enter. This should probably be a preference for those who are relatively perfect typists.
  - I would imagine this technically enables more intuitive click modifiers like Cmd + Click instead of shortcut keys like the previous Homerow app.
    - I hope to see combination modifiers, e.g., Shift + Cmd + Click in the future.

**What I don't particularly like**

- Having to hold Shift to access Jumpers. I think there's a better way that involves Preferences to help customize the hybrid workflow...see below.

**Homerow individualization suggestions**

**Offer Jumper customization in preferences.** Jumpers are necessary for some apps and pages that contain unlabeled elements or many similarly labeled elements, regardless of workflow preference. Therefore, I recommend guiding users to select one of the following *Find modes* (or whatever mode label makes sense):

  - Find mode 1: Show Jumpers for all elements
    - Mode-specific option: Auto-click upon typing Jumper (`unchecked` allows users to retain "Press Return to Click" functionality for click safety; `checked` allows users to prefer speed over safety).
      - Enabling this mode-specific option would essentially restore the previous Homerow app's functionality.
      - Can you retain intuitive Shift and Cmd click modifiers in either mode, or would you need an option for the legacy app's click modifier workflow, as well?
  - Find mode 2: Search elements, show Jumpers intelligently
    - Use the new app's search mode that allows one to type what's already on the screen (I see much value in typing what's already on the screen when possible, i.e., no content-obscuring Jumpers unless necessary; see next thought).
    - Jumpers display only when:
      - Elements have empty descriptions and values.
      - The first 2-3 characters of an element's description or value are *not unique*.
        - Restricting "unique identification" to the first 2-3 chars helps minimize typing to locate specific elements.
          - Need to consider whether to focus on description, value, or a smart combination of both.
          - I think this restriction enables merging the best of search and Jumper modes, and would enable showing a narrowed set of Jumpers after a user types to search
            - Imagine a media app showing a playlist with a bunch of Play elements on the screen. Each Play element has a hidden description like 'Play Music for Programming 1' and 'Play Music for Programming 2'. One can type "Pla" to search, and then Homerow has enough information to show a final 1-2 letter Jumper to jump to the specific element. User then presses Return to click.
            - Of course, other Jumpers may need to exist on the screen for searching, so Homerow would ideally display Jumpers for unlabeled elements after determining non-unique labeled (semi-duplicate) elements. In the example above, Homework would ensure that unlabeled element Jumpers would not start with "P", thereby enabling the search preference for the "Pla"-labeled elements.
    - Because one types to search, there probably can't be an auto-click option without some sort of AI (overkill in this case). I think that's good. As I wrote above, I prefer searching and then hitting `Return` to click what's highlighted.
    - I think this mode would be a good default for new Homerow users (those not upgrading from a previous version).

I hope that's useful feedback. I'd be happy to discuss further in a meeting. In the meantime, thank you for your efforts!

**Comment by dexterleng** [comment]
## Jumpers Only Workflow (Preview)

No need to hold Shift to type the jumper labels :). 

You still need to press Enter to click the element though, I prefer that over the prefix workflow in old Homerow because:
1. Modifiers+Enter is symmetric to Modifier+Click2. Same thing for double click (press Enter twice in quick succession)
2. Confirmation so you don't click the wrong element3.3. 
3. Consistent with the search workflow

![CleanShot 2022-11-20 at 17 53 25](https://user-images.githubusercontent.com/34204380/202895743-22c66856-9a04-4e6e-a663-679c21e6f51f.png)

## Download

[homerow-jumpers-2-preview.zip](https://github.com/dexterleng/homerow/files/10049324/homerow-jumpers-2-preview.zip)

**Comment by brettflorio** [comment]
> No need to hold Shift to type the jumper labels :).

Thanks for adding that feature! Works well! And though I initially thought needing to hit enter would be noticeable, it doesn't bother me at all.

Not sure if you want reports like this on this ticket or a new one, but in comparing both the new and old Homerows, it seems like the new is a bit more "aggressive". On a Google Form date picker UI, for instance, the old version has jumpers for the elements I'd expect, while the new has lots of overlapping labels. In particular, the calendar isn't jumpable in New Homerow because the working jumper is under the `CX`.

<img width="208" alt="image" src="https://user-images.githubusercontent.com/34131/202998385-ba293935-c024-4b0c-bfc5-67bb439ee1f5.png">

<img width="245" alt="image" src="https://user-images.githubusercontent.com/34131/202998278-10e4bea3-febf-4638-a127-f37e18406508.png">

<img width="228" alt="image" src="https://user-images.githubusercontent.com/34131/202998345-a55ff6d5-b7ac-4789-bd1c-1582efaf9819.png">

**Comment by tadassce** [comment]
Thanks @dexterleng for bringing back the labels functionality ❤️ I'm happy to have it back as it works way faster and better for my environment/use cases.

I'd like to +1 to most things that @brettflorio mentioned:
* The additional enter, I also thought that would be an annoying extra press but in reality it doesn't bother me at all
* The overlapping labels sometimes make it harder or impossible to see what are the letters under another label
* The mouse being in the location that it clicked does get in the way, and I'd also prefer to have the mouse pointer stay in the original location if possible

And a minor esthetic thing, the new yellow is a bit jarring and I'd personally prefer something easier on the eyes.

Regardless of any of the above, thanks again, I'm happy to have this in my life 🙂🙇

**Comment by banjerluke** [comment]
Came here to complain about how the new Homerow was nearly useless to me... leaving here with a new alpha version (homerow-jumpers-5-preview.zip) that is the best version of Homerow I've ever used! 🙏

I actually appreciate the additional enter since it lets me confirm what I'm doing – I'm not the world's most accurate typer so this will save me some mis-clicks for sure! I also like being able to type a jumper label and then use the arrow keys to move around from there.

**Comment by dexterleng** [comment]
@brettflorio @tadassce It's a trade off between false-positives (duplicate labelling or labelling stuff you'd never want to click) or false-negatives (no labels shown for things you'd want to click)

Labelling is more aggressive because labels are shown for static text which wasn't present in Vimac/old Homerow.

An example of a case where you would want more aggressive hinting is in Twitter - there's no way to click on the Tweet card without clicking on the text:

![CleanShot 2022-11-30 at 13 29 20](https://user-images.githubusercontent.com/34204380/204715187-d8e5f43c-8798-43ec-ad95-88e510422184.png)

Personally I prefer more stuff to be hinted and then not being able to click something and having to use the mouse.

As for overlapping labels, yeah, there's two ways to go about it:
1. Typing the label on-top, which reveals the bottom label, then typing the bottom label
2. Typing the label on-top or to the side, then using arrow keys to navigate to the correct one.

**Comment by dexterleng** [comment]
Anyways, just shipped Labels in Alpha 9 🚀

**Comment by banjerluke** [comment]
Related to having to press Enter to tap a label: I would love to have an option to tap spacebar instead of enter (or some other key) to click and keep labels visible so that I can click again.

**Comment by brettflorio** [comment]
> 2. Typing the label on-top or to the side, then using arrow keys to navigate to the correct one.

@dexterleng I didn't know about that functionality until @banjerluke mentioned it. That works well, and it's great that tab/shift+tab works too. (Easier to hit than the arrows.)

**Comment by dexterleng** [comment]
Closing this now. Thanks everyone for giving the previews a try. 🙏
